### ISSUE TITLE
🚀 Feature: Filter duplicate techdocs entries from search

### ISSUE THREAD

**Issue by ovalice** [backstage/backstage#24918]
### 🔖 Feature description

I would like to be able to filter and remove duplicate entries in Backstage search. When you provide a search term to search, I will often receive duplicate entries of the same techdoc:

![image](https://github.com/backstage/backstage/assets/31350208/5c9dc88c-2bcd-4ad3-991f-b6a8d55b403a)

It looks like the default behavior of search is that it displays the techdoc itself, and then it can appear again for each header in the document that also matches the search query. Note that it doesn't have to be the header itself that has to match search query. The header will also appear in the search results when the content that is associated with that header matches the search query.

### 🎤 Context

When using search to find techdocs, I notice that I often get documents in my view that refer to the same document, with the difference being that the underlying links refer to different locations within the document. This mucks up the search results with lots of the same documents, and that's a problem when what I need is a lot further down. I would like to ensure that techdocs only appear once in the results, so I get a better overview. I couldn't find any documentation of a feature that provides this functionality right now.

### ✌️ Possible Implementation

For my use case, I only need to filter techdocs results, so that would be sufficient. I'm not sure if it makes sense to implement this use case for other categories as well.

I'm nearly clueless about how search works, so I'll have to dive in. Perhaps we can pass an option to `<TechDocsSearchResultListItem />`, though I think filtering should preferably be done in the backend (I have no idea how that works right now). I'm also worried about how this should work in conjunction with pagination so that the next page won't show results of the previous page, but perhaps this shouldn't be a blocker.

Perhaps we could introduce an option in the search UI to filter duplicates/ignore URLs with anchors, although I would like to be able to set a default that documents get filtered.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by ovalice** [comment]
Hmm, I've been looking into it now, but this seems tricky to implement if we want pagination to work as expected. Because most of the pagination heavy lifting is done by each individual search engine, support for filtering would need to be implemented for each one of them in order to fulfill the page size that is requested.

Filtering on the front-end is also possible, but then we potentially end up with less documents than the page size in the view. Which isn't the end of the world, but not ideal.

**Comment by ovalice** [comment]
I found out I misconfigured Search, which caused results to be significantly worse results from using the `LunrSearchEngine` instead of the `PgSearchEngine`. 
After fixing this, the duplicate results don't seem like a huge detriment anymore. I've decided to close this issue, both because it's not a huge problem to me anymore, and due to the relatively high amount of effort needed to resolve a not so significant problem. I'm happy to re-open this though if there are people that still think this is a problem.

**Comment by ovalice** [comment]
I found out I misconfigured Search, which caused results to be significantly worse results from using the `LunrSearchEngine` instead of the `PgSearchEngine`.

After fixing this, the duplicate results don't seem like a huge detriment anymore. I've decided to close this issue, both because it's not a huge problem to me anymore, and due to the relatively high amount of effort needed to resolve a not so significant problem. I'm happy to re-open this though if there are people that still think this is a problem.

**Comment by slarwise** [comment]
@tenstan Do you have any details of how you misconfigured Search? Also running into this problem with duplicate search entries.

**Comment by tenstan** [comment]
@slarwise My misconfiguration at the time was the use of the in-memory [Lunr](https://backstage.io/docs/features/search/search-engines/#lunr) search engine. It was supposed to be using the Postgres search engine, but I was missing some config. I have since changed jobs and don't have access to the original source anymore, but I'm pretty sure I was only missing the config as described [in the docs](https://backstage.io/docs/features/search/search-engines/#postgres).

The Postgres search engine seems to be a lot better at showing relevant results at the top of the page. I'm assuming the duplicate search results are still appearing somewhere within the search results, but in a different order, so they are less visible. That's why the duplicate search results weren't a problem for me anymore. It's been some time ago since I've created this issue though, so I might've forgotten some details.
